processor/groupbyattrsprocessor/processor.go

```diff
@@ -0,0 +1,146 @@
+)
+
+type groupbyattrsprocessor struct {
+	logger      *zap.Logger
+	groupByKeys []string
+}
+
+// ProcessTraces process traces and groups traces by attribute.
+func (gap *groupbyattrsprocessor) ProcessTraces(_ context.Context, td pdata.Traces) (pdata.Traces, error) {
+	rss := td.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		if rs.IsNil() {
+			continue
+		}
+
+		extractedResourceSpans := newSpanAttributeGroups()
+
+		ilss := rs.InstrumentationLibrarySpans()
+		for j := 0; j < ilss.Len(); j++ {
+			ils := ilss.At(j)
+			for k := 0; k < ils.Spans().Len(); k++ {
+				span := ils.Spans().At(k)
+
+				groupedAnything, groupedAttrMap, nonGroupedAttrMap := gap.splitAttrMap(span.Attributes())
+				if groupedAnything {
+					mNumGroupedSpans.M(1)
+					overwriteAttributes(nonGroupedAttrMap, span.Attributes())
+				} else {
+					mNumNonGroupedSpans.M(1)
+				}
+
+				entry := extractedResourceSpans.attributeEntry(groupedAttrMap, rs.Resource())
+				entry.instrumentationLibrarySpans(ils.InstrumentationLibrary()).Spans().Append(span)
+			}
+		}
+
+		for _, ers := range extractedResourceSpans.entries {
+			rs.InitEmpty()
+			ers.resourceSpans.CopyTo(rs)
+		}
+		mDistSpanGroups.M(int64(len(extractedResourceSpans.entries)))
+	}
+
+	return td, nil
+}
+
+func (gap *groupbyattrsprocessor) ProcessLogs(_ context.Context, ld pdata.Logs) (pdata.Logs, error) {
+	rl := ld.ResourceLogs()
+
+	for i := 0; i < rl.Len(); i++ {
+		ls := rl.At(i)
+		if ls.IsNil() {
+			continue
+		}
+
+		extractedResourceLogs := newLogAttributeGroups()
+
+		ills := ls.InstrumentationLibraryLogs()
+		for j := 0; j < ills.Len(); j++ {
+			ill := ills.At(j)
+			for k := 0; k < ill.Logs().Len(); k++ {
+				log := ill.Logs().At(k)
+
+				groupedAnything, groupedAttrMap, nonGroupedAttrMap := gap.splitAttrMap(log.Attributes())
+				if groupedAnything {
+					mNumGroupedLogs.M(1)
+					overwriteAttributes(nonGroupedAttrMap, log.Attributes())
+				} else {
+					mNumNonGroupedLogs.M(1)
+				}
+
+				entry := extractedResourceLogs.attributeEntry(groupedAttrMap, ls.Resource())
+				entry.instrumentationLibraryLogs(ill.InstrumentationLibrary()).Logs().Append(log)
+			}
+		}
+
+		for _, erl := range extractedResourceLogs.entries {
+			ls.InitEmpty()
+			erl.resourceLogs.CopyTo(ls)
+		}
+		mDistLogGroups.M(int64(len(extractedResourceLogs.entries)))
+	}
+
+	return ld, nil
+}
+
+func overwriteAttributes(selectedAttrs, targetAttrs pdata.AttributeMap) {
+	targetAttrs.InitEmptyWithCapacity(selectedAttrs.Len())
+	selectedAttrs.CopyTo(targetAttrs)
+}
+
+// splitAttrMap splits the AttributeMap by groupByKeys and returns three-tuple:
+//  - the first element indicates if anything was matched (true) or nothing (false)
+//  - the second element contains groupByKeys that did not match the keys
+//  - the third element contains groupByKeys that match given keys
+func (gap *groupbyattrsprocessor) splitAttrMap(attrMap pdata.AttributeMap) (bool, pdata.AttributeMap, pdata.AttributeMap) {
```

Comment:
Maybe create struct for the returned tuple?

Reply:
IMO this would be a 'Go antipattern' - to replace separate return values with a struct. At least for this use case here